src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemorySegmentPool.java

```diff
@@ -0,0 +1,201 @@
+ * can be set.
+ *
+ * <br />
+ *
+ * When the pool can't provide segment from bucket the new segment is created, there's no upper
+ * bound.
+ *
+ * <br />
+ *
+ * When the segment is returned back to pool, pool checks if the bucket size is less than maximum.
+ * In such a case segment entry is put back, otherwise it is deallocated. This check is not atomic
+ * with put (due to performance reasons), however low probable queue finally can contain
+ * more elements than max.
+ *
+ * <br />
+ * Please note: that this pool can allocate segments of larger size than requested
+ * (almost 2x as requested), if for small it can be ok, than for larger segments it can be
+ * risky. I. e. request for segment for 1MB (2^20), will result in segment of 1MB, however requesting for
+ * segment of size 1MB + 1(2^20 + 1), will result in segment of size 2^21). However in case of
+ * 1GB + 1, pool will allocate 2GB of memory.
+ */
+public class MemorySegmentPool {
+  private static final int[] DEFAULT_MAX_SIZES = new int[Long.SIZE];
+  private static final ResourceScope GLOBAL = ResourceScope.globalScope();
+
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  private final SpinLockQueue<MemoryPoolSegment> segmentsDequeue[] = new SpinLockQueue[Long.SIZE];
+
+  private final ResourceScope scope;
+
+  static {
+    int idx = 0;
+    int cores = Runtime.getRuntime().availableProcessors();
+
+    for (; idx <= 10; idx++) {
+      DEFAULT_MAX_SIZES[idx] = 256;
+    }
+
+    // Sizes up to 64kb
+    for (; idx <= 16; idx++) {
+      DEFAULT_MAX_SIZES[idx] = cores * 4;
+    }
+
+    // Sizes up to 1MB {
+    for (; idx <= 20; idx++) {
+      DEFAULT_MAX_SIZES[idx] = cores;
+    }
+
+    DEFAULT_MAX_SIZES[idx++] = Math.min(cores, 2); //2mb
+    DEFAULT_MAX_SIZES[idx++] = Math.min(cores, 1); //4mb
+    // Rest 0
+  }
+
+  public MemorySegmentPool(ResourceScope scope) {
+    this(DEFAULT_MAX_SIZES, scope);
+  }
+  public MemorySegmentPool(int maxSizes[], ResourceScope scope) {
+    this.scope = scope;
+
+    validateMaxSizes(maxSizes);
+    for (int i=0; i < segmentsDequeue.length; i++) {
+      segmentsDequeue[i] = new SpinLockQueue<>(maxSizes[i]);
+    }
+  }
+
+  public SegmentAllocator allocatorForScope(ResourceScope resourceScope) {
+    // Prevent scope managing this pool to go away, when dependant allocator is alive
+    final var handle = scope.acquire();
+    resourceScope.addOnClose(handle::close);
+    return (bytesSize, bytesAlignment) -> getSegmentForScope(resourceScope, bytesSize, bytesAlignment);
+  }
+
+  public MemoryPoolSegment getSegmentEntryByLayout(MemoryLayout layout) {
+    return getSegmentEntryBySize(layout.byteSize(), layout.byteAlignment());
+  }
+
+  /**
+   * Gets segment from pool or allocates new one. Internally segments are cached.
+   * The size of segment can be larger than requested.
+   *
+   * @param size the size of segment.
+   *
+   * @return segment of size at least `size`
+   */
+  @ForceInline
```

Comment:
I've tried removing @ForceInline here and I did not notice any regression in the benchmark.